### PR TITLE
ERM-83: Manage external data sources

### PR DESCRIPTION
**Pull request by aditya-matukumalli**
- Setting page for managing Local KB Admin external data source configurations
- New card based UI is used to render the data sources
- Used React final form. 

image

Yet to do:

- Tests

### CODE REVIEW

File: src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesEdit.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    meta: PropTypes.shape({
+      invalid: PropTypes.bool,
+      pristine: PropTypes.bool,
+      submitting: PropTypes.bool,
+    })
+  }
+
+  render() {
+    const { input: { value, name } } = this.props;
+    const kbadapterTypes = [{ value: 'org.olf.kb.adapters.EbscoKBAdapter', label: 'org.olf.kb.adapters.EbscoKBAdapter' }];
```
Comment (doytch) [inline src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesEdit.js]: Can you just define these as member variables or constants? No need to recreate this array and object each render.

File: src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesEdit.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    const recordTypes = [{ value: '1', label: 'Package' }];
```
Comment (doytch) [inline src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesEdit.js]: Can you just define these as member variables or constants? No need to recreate this array and object each render.

File: src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesEdit.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        data-test-external-data-source-edit
+        headerStart={(
+          <strong>
+            {value.id ?
+              <FormattedMessage id="ui-local-kb-admin.settings.externalDataSources.editExternalKb" />
+              :
+              <FormattedMessage id="ui-local-kb-admin.settings.externalDataSources.newExternalKb" />}
+          </strong>
+        )}
+        headerEnd={this.props.actionButtons}
```
Comment (doytch) [inline src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesEdit.js]: If you're destructuring props, destructure these as well. (I now see I made this error in mine as well.)

File: src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesEdit.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+      >
+        <Row>
+          <Col xs={4} md={4}>
```
Comment (doytch) [inline src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesEdit.js]: No need to specify `xs` and `md` if they're the same. Same throughout.

File: src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesEdit.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+              validate={required}
+            />
+          </Col>
+        </Row>
+        <Field
+          component={TextField}
+          label={<FormattedMessage id="ui-local-kb-admin.settings.externalDataSources.uri" />}
+          name={`${name}.uri`}
+        />
+        <Layout className="padding-bottom-gutter">
```
Comment (doytch) [inline src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesEdit.js]: Is this because of specific `Checkbox` rendering? Curious...

File: src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesFields.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        id: PropTypes.string,
+      }).isRequired,
+    }).isRequired,
+    meta: PropTypes.shape({
+      invalid: PropTypes.bool,
+      pristine: PropTypes.bool,
+      submitting: PropTypes.bool,
+    }),
+    mutators: PropTypes.shape({
+      resetTermState: PropTypes.func,
```
Comment (doytch) [inline src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesFields.js]: What are these _terms_ you speak of? 🤔

File: src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesForm.js [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        <Callout ref={ref => { this.callout = ref; }} />
+      </Pane>
+    );
+  }
+}
+export default stripesFinalForm({
+  enableReinitialize: true,
+  keepDirtyOnReinitialize: false,
+  mutators: {
+    resetTermState: (args, state, tools) => {
```
Comment (doytch) [inline src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesForm.js]: _Terms_, eh? 🤔

Comment (doytch) [reply src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesForm.js]: Also, the term state resetting doesn't really do anything, does it? I removed it from mine.

File: src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesView.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    })
+  }
+
+  render() {
+    const { input: { value } } = this.props;
+
+    return (
+      <Card
+        data-test-external-data-source-view
+        headerStart={<strong>External Kb</strong>}
```
Comment (doytch) [inline src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesView.js]: i18n

File: src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesView.js [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        headerEnd={this.props.actionButtons}
+      >
+        <Row>
+          <Col xs={3} md={3}>
+            <KeyValue
+              data-test-external-data-source-name="bar"
```
Comment (doytch) [inline src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesView.js]: `bar`?

File: src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesView.js [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    const { input: { value } } = this.props;
+
+    return (
+      <Card
+        data-test-external-data-source-view
+        headerStart={<strong>External Kb</strong>}
+        headerEnd={this.props.actionButtons}
+      >
+        <Row>
+          <Col xs={3} md={3}>
```
Comment (doytch) [inline src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesView.js]: Run through this whole PR looking for unnecessary setting of `md`.

File: test/bigtest/interactors/external-data-sources.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  editRecordType = selectable('[data-test-external-data-source-record-type-edit]');
+}
+
+@interactor class ExternalDataSourcesListInteractor {
+  size = count('[data-test-external-data-source-view]');
+  items = collection('[data-test-external-data-source-view]', ExternalDataSourceViewInteractor);
+  itemsEdit = collection('[data-test-external-data-source-edit]', ExternalDataSourceEditInteractor);
+}
+
+export default @interactor class ExternalDataSourcesInteractor {
```
Comment (doytch) [inline test/bigtest/interactors/external-data-sources.js]: Nice hierarchy and scoping with this interactor, it's good!

File: test/bigtest/tests/settings-externalDataSources-test.js [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+      await externaldatasources.externalDataSourceList.itemsEdit(0).clickSaveButton();
+    });
+
+    describe('creating a new data source', () => {
+      it('sholud render the expected source', () => {
+        expect(externaldatasources.externalDataSourceList.items(0).name.value.text).to.equal(name);
+      });
+
+      describe('editing the source', () => {
+        const editedName = 'aa edited name';
```
Comment (doytch) [inline test/bigtest/tests/settings-externalDataSources-test.js]: Could you edit some more properties? `Name` is always fairly trivial since it's a textbox but some of the other types (checkboxes, etc) have edge cases that can crop up.

File: test/bigtest/tests/settings-externalDataSources-test.js [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+      this.visit('/settings/local-kb-admin/external-data-sources');
+      await externaldatasources.clickNewButton();
+      await externaldatasources.externalDataSourceList.itemsEdit(0).editName(name);
+      await externaldatasources.externalDataSourceList.itemsEdit(0).editType(type);
+      await externaldatasources.externalDataSourceList.itemsEdit(0).editRecordType(recordType);
+      await externaldatasources.externalDataSourceList.itemsEdit(0).clickSaveButton();
+    });
+
+    describe('creating a new data source', () => {
+      it('sholud render the expected source', () => {
```
Comment (doytch) [inline test/bigtest/tests/settings-externalDataSources-test.js]: tpyo

File: test/bigtest/tests/settings-externalDataSources-test.js [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+        expect(externaldatasources.externalDataSourceList.items(0).name.value.text).to.equal(name);
+      });
+
+      describe('editing the source', () => {
```
Comment (doytch) [inline test/bigtest/tests/settings-externalDataSources-test.js]: Can you add a case for maybe cancelling out of an edit to ensure the previous values got restored?

File: src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesEdit.js (hunk 5 above)
Comment (aditya-matukumalli) [reply src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesEdit.js]: Yes, if not they look like this:

image

File: src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesFields.js (hunk 6 above)
Comment (aditya-matukumalli) [reply src/settings/components/ExternalDataSourcesConfig/ExternalDataSourcesFields.js]: artifact

File: test/bigtest/tests/settings-externalDataSources-test.js (hunk 12 above)
Comment (aditya-matukumalli) [reply test/bigtest/tests/settings-externalDataSources-test.js]: Added additional tests for testing the checkboxes status after editing and saving. The other inputs here are either textFields or Select dropdowns (with just one option), so ignored the other inputs for now.

File: test/bigtest/tests/settings-externalDataSources-test.js (hunk 14 above)
Comment (aditya-matukumalli) [reply test/bigtest/tests/settings-externalDataSources-test.js]: done

File: src/settings/routes/ExternalDataSourcesSettingsRoute.js [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+      externalKbs: PropTypes.shape({
+        DELETE: PropTypes.func.isRequired,
+        POST: PropTypes.func.isRequired,
+        PUT: PropTypes.func.isRequired,
+      }),
+    }),
+  };
+
+  static manifest = Object.freeze({
+    externalKbs: {
```
Comment (doytch) [inline src/settings/routes/ExternalDataSourcesSettingsRoute.js]: Can you sort this by `id;desc` so that when you create a new data source, it'll stay at the top of the list and not jump down to the bottom?
